### PR TITLE
Async pure OCaml TLS, client-side only

### PR DESCRIPTION
**Pull request by mbacarella**
Adds support for tls-async, client-side only.

Depends on https://github.com/anmonteiro/gluten/pull/31

### CODE REVIEW

**Comment by mbacarella** [comment]
Oh right, I also refreshed the certificates directory. My openssl implementation complained that 1024 bit keys weren't supported anymore so these were created with 2048 bits.

File: examples/alpn/lib/http1_handler.ml [hunk 1]
```diff
@@ -12,12 +12,12 @@ let redirect_handler : Unix.sockaddr -> Reqd.t Gluten.reqd -> unit =
   Reqd.respond_with_string reqd response ""
 
 let redirect_error_handler
-    :  Unix.sockaddr -> ?request:Request.t -> _ -> (Headers.t -> Body.Writer.t)
-    -> unit
+    :  Unix.sockaddr -> ?request:Request.t -> _
```
Comment (anmonteiro) [inline examples/alpn/lib/http1_handler.ml:15]: Please revert this file. This is supposed to be http/af, not h2, as it is an HTTP/1.x handler.

File: examples/async/tls_async_https_echo_client_post.ml [hunk 2]
```diff
@@ -0,0 +1,86 @@
+open Core
```
Comment (anmonteiro) [inline examples/async/tls_async_https_echo_client_post.ml:1]: could you keep the naming consistent with the other example? `async_tls_https_echo_client_post.ml`

File: examples/lwt/lwt_h2c.ml [hunk 3]
```diff
@@ -105,8 +105,8 @@ let connection_handler =
         Status.to_string error
     in
     let body = handle Headers.empty in
-    Body.Writer.write_string body message;
-    Body.Writer.close body
+    H2.Body.Writer.write_string body message;
```
Comment (anmonteiro) [inline examples/lwt/lwt_h2c.ml:108]: Please revert this file, it's supposed to be `Httpaf.Body`.

**Comment by anmonteiro** [comment]
If you merge master onto this branch, CI will probably be happy.

File: examples/lwt/lwt_h2c.ml (hunk 3 above)
Comment (mbacarella) [reply examples/lwt/lwt_h2c.ml]: done

File: examples/async/tls_async_https_echo_client_post.ml (hunk 2 above)
Comment (mbacarella) [reply examples/async/tls_async_https_echo_client_post.ml]: done

File: examples/alpn/lib/http1_handler.ml (hunk 1 above)
Comment (mbacarella) [reply examples/alpn/lib/http1_handler.ml]: done

**Comment by mbacarella** [comment]
> If you merge master onto this branch, CI will probably be happy.

I think it was already caught up to master?

File: async/h2_async.ml [hunk 4]
```diff
@@ -175,4 +175,51 @@ module Client = struct
...
+        ?push_handler
+        ~error_handler
+        socket
+        where_to_connect
+      =
+      Client_runtime.create_default
+        ~alpn_protocols:[ "http/1.1" ]
+        socket
+        where_to_connect
+      >>= fun ssl_client ->
```
Comment (anmonteiro) [inline async/h2_async.ml:217]:
```suggestion
      >>= fun tls_client ->
```

**Review by anmonteiro — APPROVED** [review]
very cool, thanks!

File: async/h2_async.ml [hunk 5]
```diff
@@ -175,4 +175,51 @@ module Client = struct
...
+        ~error_handler
+        socket
+        where_to_connect
+      =
+      Client_runtime.create_default
+        ~alpn_protocols:[ "http/1.1" ]
+        socket
+        where_to_connect
+      >>= fun tls_client ->
+      create_connection ~config ?push_handler ~error_handler ssl_client
```
Comment (anmonteiro) [inline async/h2_async.ml:218]:
```suggestion
      create_connection ~config ?push_handler ~error_handler tls_client
```

**Pull request merged**